File: Data Structures & Algorithms/surrounded-regions/submission-0.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<char>>& board) {
        if( row < 0 || row == board.size() || col < 0 || col == board[row].size() || board[row][col] != 'O') {
            return;
        }
        board[row][col] = '#';
        dfs(row + 1, col, board);
        dfs(row - 1, col, board);
        dfs(row, col + 1, board);
        dfs(row, col -1, board);
    }

    void solve(vector<vector<char>>& board) {
        for(int i=0;i< board.size();i++) {
            for(int j = 0;j<board[i].size();j++){
                if(board[i][j] == 'O' && (i == 0 || i == board.size()-1) || (j == 0 || j == board[i].size()-1)){
                 dfs(i, j, board);       
                }
            }  
        }

        for(int i=0;i<board.size();i++) {
            for(int j = 0;j<board[i].size();j++){
                if(board[i][j] == 'O'){
                 board[i][j] = 'X';       
                }
            }
            
        }

        for(int i=0;i<board.size();i++) {
            for(int j = 0;j<board[i].size();j++){
                if(board[i][j] == '#'){
                 board[i][j] = 'O';       
                }
            }
            
        }
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        rows, cols = len(grid), len(grid[0])
        visited = set()
        islands = 0

        def bfs(r,c):
            q = deque()
            visited.add((r,c))
            q.append((r,c))
            while q:
                row, col = q.popleft()
                paths = [[1,0], [0,1], [-1,0], [0,-1]]
                for pr,pc in paths:
                    if (row + pr) in range(rows) and (col + pc) in range(cols) and grid[row + pr][col + pc] == "1" and ((row + pr), (col + pc)) not in visited:
                        q.append((row+pr, col+pc))
                        visited.add((row+pr, col+pc))
                
               


        for r in range(rows):
            for c in range(cols):
                if (r,c) not in visited and grid[r][c] == "1":
                    bfs(r,c)
                    islands += 1
        return islands
